### PR TITLE
Thiserror

### PR DESCRIPTION
**Pull request by Aloso**
This adds the `thiserror` crate and updates error handling code to use error enums.

### CODE REVIEW

File: src/params.rs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+use std::collections::BTreeMap;
+
+/// A key/value collection containing strings.
+/// This is used for mediawiki requests.
+pub type Params = BTreeMap<String, String>;
```
Comment (maroider) [inline src/params.rs]: What's the rationale for using a `BTreeMap` instead of a `HashMap`?

File: src/lib.rs [hunk 2]
```diff
@@ -9,24 +7,29 @@ extern crate hmac;
...
 pub mod page;
+pub mod params;
 pub mod title;
 pub mod user;
+
+pub use method::Method;
+pub use page::Page;
+pub use params::Params;
+pub use title::Title;
+pub use user::User;
```
Comment (maroider) [inline src/lib.rs]: I'm not sure I'm a fan of re-exporting stuff here. The modules in question certainly have a tiny surface each, so I can see the appeal of doing it, but I'm not entirely convinced.

File: src/api.rs [hunk 3]
```diff
@@ -460,12 +455,7 @@ impl Api {
 
     /// Runs a query against the MediaWiki API, using `method` GET or POST.
     /// Parameters are a hashmap; `format=json` is enforced.
-    async fn query_api_json_mut(
-        &mut self,
-        params: &Params,
-        method: Method,
-    ) -> Result<Value, Error> {
-        let mut params = params.clone();
+    async fn query_api_json_mut(&mut self, mut params: Params, method: Method) -> Result<Value> {
```
Comment (maroider) [inline src/api.rs:454]: I like how you've shifted the burden of `.clone()`-ing onto the caller here.

File: src/params.rs (hunk 1 above)
Comment (Aloso) [reply src/params.rs]: @maroider It's more memory efficient. I also heard that it's faster than a HashMap, if the number of entries is small, although I haven't benchmarked it.

EDIT: I was mistaken: Even for collections with 10 entries, lookup in a HashMap is almost twice as fast as in a BTreeMap. However, inserting an entry is faster for BTreeMaps. I don't know about the memory footprint, not sure how I could benchmark it.

File: src/lib.rs (hunk 2 above)
Comment (Aloso) [reply src/lib.rs]: I'll remove it then.

File: src/api.rs (hunk 3 above)
Comment (Aloso) [reply src/api.rs:454]: Thanks! The downside is that the signatures are no longer consistent (some functions accept `&Params` and some accept `Params`).

File: src/api.rs [hunk 4]
```diff
@@ -14,16 +14,14 @@ The `Api` class serves as a univeral interface to a MediaWiki API.
...
-use crate::title::Title;
-use crate::{method::Method, user::User, Params};
+use crate::{
+    error::{Error, Result},
+    hmac::{Mac, NewMac},
+    method::Method,
+    params::Params,
+    title::Title,
+    user::User,
+};
```
Comment (maroider) [inline src/api.rs]: We should try to be consistent with the import style that's already in place (one `use` per module).

**Review by maroider — DISMISSED** [review]

File: src/params.rs (hunk 1 above)
Comment (maroider) [reply src/params.rs]: I think we should revert to using a `HashMap` and then come back to optimization later.

**Comment by Aloso** [comment]
@maroider thanks for the review!

**Pull request merged**